File: link-list/longest-streak/longest-streak.go
// Write a function, longestStreak, that takes in the head of a linked list as an argument. The function should return the length of the longest consecutive streak of the same value within the list.
package main

import "fmt"

type Node struct {
	value int
	next  *Node
}

func NewNode(value int) *Node {
	return &Node{value: value}
}

func printLinkedList(head *Node) {
	if head == nil {
		return
	}
	fmt.Println(head.value)
	printLinkedList(head.next)
}

func longestStreak(head *Node) int {
	if head == nil {
		return 0
	}
	maxStreak := 1
	curStreak := 1
	curValue := head.value
	pointer := head.next
	for pointer != nil {
		if curValue == pointer.value {
			curStreak++
		} else {
			curStreak = 1
			curValue = pointer.value
		}
		if curStreak > maxStreak {
			maxStreak = curStreak
		}
		pointer = pointer.next
	}

	return maxStreak
}

func main() {
	a := NewNode(5)
	b := NewNode(5)
	c := NewNode(7)
	d := NewNode(7)
	e := NewNode(7)
	f := NewNode(6)

	a.next = b
	b.next = c
	c.next = d
	d.next = e
	e.next = f

	x := NewNode(5)

	fmt.Println(longestStreak(a))
	fmt.Println(longestStreak(x))
	fmt.Println(longestStreak(nil))
}

File: link-list/longest-streak/longest-streak.js
// Write a function, longestStreak, that takes in the head of a linked list as an argument. The function should return the length of the longest consecutive streak of the same value within the list.

class Node {
  constructor(val) {
    this.val = val;
    this.next = null;
  }
}

const longestStreak = head => {
    if(head === null) return 0
    let maxStreak = 1
    let curStreak = 1
    let curValue = head.val
    let pointer = head.next
    while (pointer !== null) {
        if (curValue === pointer.val) {
            curStreak++
        } else {
            curStreak = 1
            curValue = pointer.val
        }
        maxStreak = curStreak > maxStreak ? curStreak : maxStreak;
        pointer = pointer.next
    }
    return  maxStreak
};

const a = new Node(5);
const b = new Node(5);
const c = new Node(7);
const d = new Node(7);
const e = new Node(7);
const f = new Node(6);

a.next = b;
b.next = c;
c.next = d;
d.next = e;
e.next = f;

console.log(longestStreak(a)); // 3